include/kframework/lemmas/evm-int-simplification.k

```diff
@@ -0,0 +1,55 @@
+requires "evm-types.md"
+
+module EVM-INT-SIMPLIFICATION
+    imports EVM-INT-SIMPLIFICATION-HASKELL
+endmodule
+
+module EVM-INT-SIMPLIFICATION-HASKELL [symbolic, kore]
+    imports EVM-INT-SIMPLIFICATION-COMMON
+endmodule
+
+module EVM-INT-SIMPLIFICATION-COMMON
+    imports INT
+    imports BOOL
+    imports EVM-TYPES
+
+  // ###########################################################################
+  // up/Int
+  // ###########################################################################
+
+    rule [upInt-lt-true]:
+      ((X up/Int Y) *Int Y) <Int Z => true
+      requires X +Int Y <=Int Z andBool 0 <Int Y
+      [simplification]
+
+    rule [upInt-lt-false]:
+      ((X up/Int Y) *Int Y) <Int Z => false
+      requires Z <=Int X andBool 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-leq]:
+      ((X up/Int Y) *Int Y) <=Int X => X modInt Y ==Int 0
+      requires 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-gt]:
+      X <Int ((X up/Int Y) *Int Y) => X modInt Y =/=Int 0
+      requires 0 <Int Y
+      [simplification]
```

Comment:
I still don't really like these lemmas. Are we sure they're needed?

Could we also replace them with lemmas like this:

```
rule ((X up/Int Y) *Int Y) => X +Int (X modInt Y) requires 0 <Int Y andBool 0 <=Int X [simplification]
```

This I feel expresses the semantics of `_up/Int_` better, while being more general (so that we can have general lemmas about `_<=Int_` and `_<Int_` without `_up/Int_`).

Reply:
I'm not sure that holds... take, for example: `Y ==Int 5` and `X ==Int 6, 7, 8, 9`. We have that `(X up/Int Y) *Int Y ==Int 10` for all of them, but `X +Int (X modInt Y) ==Int 7, 9, 11, 13`, right?

By definition, I think `X up/Int Y ==Int (X +Int Y -Int 1) /Int Y`?

The lemmas above aren't necessarily needed, I just have all six `<=, <, >, >=, ==, =/=` for completeness. I think @lucasmt ran into the `<Int`  version only?

Reply:
Yes you're right, not sure what I was thinking. I would prefer lemma which directly represented the semantics of `up/Int` better, but I guess we will have to live with ones that define it relative to eqaulity instead.